### PR TITLE
Updated part of the docstrings of the server

### PR DESCRIPTION
**Pull request by bartvanb**
**Work-in-progress** let's merge this when all server docstrings are updated

### CODE REVIEW

File: vantage6-server/vantage6/server/permission.py [hunk 1]
```diff
@@ -18,27 +19,46 @@
...
+            Scope within which to apply the rule
+        operation: Operation
+            What operation the rule applies to
+        """
         permission = Permission(RuleNeed(self.name, scope, operation))
         self.__setattr__(f'{operation.value}_{scope.value}', permission)
 
-    def get(self, scope: Scope, operation: Operation):
-        return self.__getattribute__(f'{scope}_{operation}')
-
```
Comment (bartvanb) [inline vantage6-server/vantage6/server/permission.py:32]: Note: we are deleting an unused function here!

File: vantage6-server/vantage6/server/model/user.py [hunk 2]
```diff
@@ -139,28 +179,87 @@ def is_blocked(self, max_failed_attempts: int,
...
+        NoResultFound
+            If no user with the given email exists
+        """
         session = DatabaseSessionManager.get_session()
         return session.query(cls).filter_by(email=email).one()
 
     @classmethod
-    def get_user_list(cls, filters=None):
-        session = DatabaseSessionManager.get_session()
-        return session.query(cls).all()
```
Comment (bartvanb) [inline vantage6-server/vantage6/server/model/user.py:154]: Note: we are deleting an unused function here!

**Comment by bartvanb** [comment]
Let's see if we can merge this now - updating all server docstrings at once will only keep creating a lot of merge conflicts

File: vantage6-server/vantage6/server/permission.py [hunk 3]
```diff
@@ -18,27 +19,46 @@
...
+    Class that tracks a set of all rules for a certain resource name
 
-    def __init__(self, name):
+    Parameters
+    ----------
+    name: str
+        Name of the resource endpoint (e.g. node, organization, user)
+    """
+
+    def __init__(self, name) -> None:
```
[the remark this answers is not in the document]
Comment (frankcorneliusmartin) [reply vantage6-server/vantage6/server/permission.py:31]: Shouldnt we keep `RuleCollection` free from resources? Now it looks like (looks like) they need to be together (which is currently the case) but this is not strictly necessary. 

In the function `assign_rule_to_node` you changed all to resource, i would prefer to keep this as a collection (and this happens to be a resource collection). 

Buttttt, i also can live with this.

File: vantage6-server/vantage6/server/permission.py [hunk 4]
```diff
@@ -158,7 +250,7 @@ def __getattr__(self, name: str):
             raise e
 
     @staticmethod
-    def rule_exists_in_db(name, scope, operation):
+    def rule_exists_in_db(name, scope, operation) -> bool:
```
Comment (frankcorneliusmartin) [inline vantage6-server/vantage6/server/permission.py:253]: Typehints are missing

File: vantage6-server/vantage6/server/permission.py [hunk 5]
```diff
@@ -184,7 +276,20 @@ def rule_exists_in_db(name, scope, operation):
...
+        Check if a user has all of a set of rules
+
+        Parameters
+        ----------
+        rules: List[Rule]
+            List of rules that user is checked to have
+
+        Returns
+        -------
+        Union[dict, bool]
```
Comment (frankcorneliusmartin) [inline vantage6-server/vantage6/server/permission.py:290]: Gehehe:  `|` looks so much better. Although this isn't wrong ofc.

File: vantage6-server/vantage6/server/resource/__init__.py [hunk 6]
```diff
@@ -68,16 +144,36 @@ def obtain_organization_id():
...
+        db.Organization
+            Organization model
+        """
         return db.Organization.get(cls.obtain_organization_id())
 
 
 # ------------------------------------------------------------------------------
 # Helper functions/decoraters ...
 # ------------------------------------------------------------------------------
 def only_for(types=['user', 'node', 'container']):
```
Comment (frankcorneliusmartin) [inline vantage6-server/vantage6/server/resource/__init__.py:162]: Typehint is missing `: list`

File: vantage6-server/vantage6/server/resource/__init__.py [hunk 7]
```diff
@@ -144,7 +252,22 @@ def get_and_update_authenticatable_info(auth_id):
 with_container = only_for(["container"])
 
 
-def parse_datetime(dt=None, default=None):
+def parse_datetime(dt=None, default=None) -> datetime:
```
Comment (frankcorneliusmartin) [inline vantage6-server/vantage6/server/resource/__init__.py:255]: Typehints are missing

File: vantage6-server/vantage6/server/resource/common/auth_helper.py [hunk 8]
```diff
@@ -15,7 +15,7 @@
 
 def user_login(
     password_policy: Dict, username: str, password: str
-) -> Tuple[Union[User, Dict], HTTPStatus]:
+) -> Tuple[User | dict, HTTPStatus]:
```
Comment (frankcorneliusmartin) [inline vantage6-server/vantage6/server/resource/common/auth_helper.py:18]: Love this

File: vantage6-server/vantage6/server/websockets.py [hunk 9]
```diff
@@ -177,46 +180,41 @@ def on_disconnect(self) -> None:
...
-                New status of the algorithm container
-            result_id : int
-                result_id for which the algorithm was running
-            collaboration_id : int
-                collaboration for which the algorithm was running
+            It should contain: 1) node_id: node_id where the algorithm
+            container was running; 2) status: New status of the algorithm
+            container; 3) result_id: result_id for which the algorithm was
+            running; and 4) collaboration_id: collaboration for which the
+            algorithm was running
```
Comment (frankcorneliusmartin) [inline vantage6-server/vantage6/server/websockets.py:217]: I understand this is due to the rendering(?). Cant we display a small example?:
```
{
    "node_id": 1,
    ...
}

**Review by frankcorneliusmartin — CHANGES_REQUESTED** [review]
That's a lot of docstrings

File: vantage6-server/vantage6/server/model/user.py (hunk 2 above)
Comment (frankcorneliusmartin) [reply vantage6-server/vantage6/server/model/user.py]: Also remote the rest in `test_models.py` L:95 - `assert User.get_user_list()`

File: vantage6-server/vantage6/server/__init__.py [hunk 10]
```diff
@@ -1,3 +1,10 @@
+"""
+The server has a central function in the vantage6 architecture. It communicates
+with the database to store which organizations, collaborations, users, etc.
```
[the remark this answers is not in the document]
Comment (bartvanb) [reply vantage6-server/vantage6/server/__init__.py:3]: done

File: vantage6-server/vantage6/server/__init__.py [hunk 11]
```diff
@@ -518,7 +535,24 @@ def __pong_response(self, node_id) -> None:
...
+               system_folders: bool = True) -> ServerApp:
+    """
+    Run a vantage6 server.
+
+    Parameters
+    ----------
+    config: str
+        Configuration file path
+    environment: str
+        Configuration environment. Default value is 'prod'.
```
[the remark this answers is not in the document]
Comment (bartvanb) [reply vantage6-server/vantage6/server/__init__.py:547]: done

File: vantage6-server/vantage6/server/model/__init__.py [hunk 12]
```diff
@@ -1,4 +1,4 @@
-from .authenticable import Authenticatable
+from .authenticatable import Authenticatable
```
[the remark this answers is not in the document]
Comment (bartvanb) [reply vantage6-server/vantage6/server/model/__init__.py:1]: done

File: vantage6-server/vantage6/server/model/organization.py [hunk 13]
```diff
@@ -69,13 +101,29 @@ def public_key(self):
             return ""
 
     @public_key.setter
-    def public_key(self, public_key_b64):
-        """Assumes that the public key is in b64-encoded."""
+    def public_key(self, public_key_b64) -> None:
```
[the remark this answers is not in the document]
Comment (bartvanb) [reply vantage6-server/vantage6/server/model/organization.py:104]: added

File: vantage6-server/vantage6/server/model/permission.py [hunk 14]
```diff
@@ -1,3 +1,13 @@
+"""
+The Permission table defines which roles have been assigned to which users.
+It can contain multiple entries for the same user if they have been assigned
+multiple roles.
+
+The Userpermission table defines which extra rules have been assigned to which
```
[the remark this answers is not in the document]
Comment (bartvanb) [reply vantage6-server/vantage6/server/model/permission.py:6]: Camelcased

File: vantage6-server/vantage6/server/model/role.py [hunk 15]
```diff
@@ -22,7 +29,21 @@ class Role(Base):
                          secondary="Permission")
 
     @classmethod
-    def get_by_name(cls, name):
+    def get_by_name(cls, name) -> Union[Role, None]:
```
[the remark this answers is not in the document]
Comment (bartvanb) [reply vantage6-server/vantage6/server/model/role.py:32]: done

File: vantage6-server/vantage6/server/model/rule.py [hunk 16]
```diff
@@ -8,21 +9,27 @@
...
     GLOBAL = "glo"
 
 
 class Rule(Base):
-    """Rules to determine permissions in an API endpoint.
+    """
+    Table that describes which rules are available.
+
+    A rule gives access to a single type of action with a given operation,
+    scope and resource on which it acts.
```
[the remark this answers is not in the document]
Comment (bartvanb) [reply vantage6-server/vantage6/server/model/rule.py:32]: nice one :-)

**Pull request merged**